runtime/vm/LayoutFFITypeHelpers.cpp

```diff
@@ -0,0 +1,229 @@
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "LayoutFFITypeHelpers.hpp"
+
+#if JAVA_SPEC_VERSION >= 16
+
+extern "C" void
+freeAllStructFFITypes(J9VMThread *currentThread, void *cifNode)
+{
+	LayoutFFITypeHelpers ffiTypeHelpers(currentThread);
+	J9JavaVM * vm = currentThread->javaVM;
+	ffi_cif *cif = (ffi_cif *)cifNode;
+	ffi_type **argTypes = cif->arg_types;
+	ffi_type *retType = cif->rtype;
+
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	if (NULL != argTypes) {
+		for (U_32 argIndex = 0; argTypes[argIndex] != NULL; argIndex++) {
+			ffiTypeHelpers.freeStructFFIType(argTypes[argIndex]);
+		}
+		j9mem_free_memory(argTypes);
+		argTypes = NULL;
+	}
+
+	if (NULL != retType) {
+		ffiTypeHelpers.freeStructFFIType(retType);
+		retType = NULL;
+	}
+}
+
+ffi_type*
+LayoutFFITypeHelpers::getArrayFFIType(char **layout, UDATA nElements) {
+	ffi_type *typeFFI = NULL;
+	ffi_type *elementType = NULL;
+	PORT_ACCESS_FROM_JAVAVM(_vm);
+
+	/* A struct starts with "#" plus the count of struct elements,
+	 * followed by the type of struct elements.
+	 * e.g. the preprocessed layout of
+	 * struct int_struct{
+	 *                   int elem1,
+	 *                   int elem2
+	 *                  }int_struct[2]
+	 *     is #2[II]
+	 */
+	if ('#' == **layout) {
+		elementType = getStructFFIType(layout);
+		if (NULL == elementType) {
+			goto done;
+		}
+	} else {
+		elementType = getPrimitiveFFIType(**layout);
+	}
+
+	if (elementType != NULL) {
+		typeFFI = (ffi_type *)j9mem_allocate_memory(sizeof(ffi_type), OMRMEM_CATEGORY_VM);
+		if (NULL == typeFFI) {
+			freeStructFFIType(elementType);
+			goto done;
+		}
+
+		typeFFI->size = 0;
+		typeFFI->alignment = 0;
+		typeFFI->type = FFI_TYPE_STRUCT;
+		typeFFI->elements = (ffi_type **) j9mem_allocate_memory(sizeof(ffi_type*) * (nElements + 1), OMRMEM_CATEGORY_VM);
+		if (NULL == typeFFI->elements) {
+			freeStructFFIType(elementType);
+			j9mem_free_memory(typeFFI);
+			typeFFI = NULL;
+			goto done;
+		}
+
+		for (U_32 elemIndex = 0; elemIndex < nElements; elemIndex++) {
+			typeFFI->elements[elemIndex] = elementType;
+		}
+		typeFFI->elements[nElements] = NULL;
+	}
+done:
+	return typeFFI;
+}
+
+ffi_type*
+LayoutFFITypeHelpers::getStructFFIType(char **layout)
+{
+	ffi_type *typeFFI = NULL;
+	/* The struct layout is incremented for the '[' case in getStructFFITypeElements(), in which case
+	 * the remaining layout string is then traversed when getStructFFITypeElements() is called recursively.
+	 */
+	(*layout) += 1; // skip over the '#' case
+	ffi_type **structElements = getStructFFITypeElements(layout);
+	PORT_ACCESS_FROM_JAVAVM(_vm);
+	typeFFI = (ffi_type *)j9mem_allocate_memory(sizeof(ffi_type), OMRMEM_CATEGORY_VM);
```

Comment:
can we allocate this out of a pool since its fixed sized?

Reply:
Will update the code accordingly to see how it goes with pool.

Reply:
I just realized we can't simply do this because the allocation of struct type is mixed with nested element struct types with different size in which case releasing the struct types aims to recursively free all memory of elements in the struct via `freeStructFFIType`, which would be complicated if allocated out of pool.

Reply:
okay, we can start with the simple solution. Can you make a new mem category for these `OMRMEM_CATEGORY_VM_FFI` this way we can track the mem usage. If its a lot we can think of ways to reduce it.

Reply:
I am on vacation from today to the end of this year. Will get back to update the code to add `OMRMEM_CATEGORY_VM_FFI` as suggested.

Reply:
`OMRMEM_CATEGORY_VM_FFI` has been added in OMR at https://github.com/eclipse/omr/pull/6282 and updated here accordingly.

Reply:
Replaced with `J9MEM_CATEGORY_VM_FFI` and revert the change in OMR at https://github.com/eclipse/omr/pull/6285 as suggested by @pshipton given our implementation in downcall & upcall is irrelevant to OMR.